types/vote.go

```diff
@@ -109,6 +109,37 @@ func (vote *Vote) CommitSig() CommitSig {
 	}
 }
 
+// ExtendedCommitSig attempts to construct an ExtendedCommitSig from this vote.
+// Panics if either the vote extension signature is missing or if the block ID
+// is not either empty or complete.
+func (vote *Vote) ExtendedCommitSig() ExtendedCommitSig {
+	if vote == nil {
+		return NewExtendedCommitSigAbsent()
+	}
+
+	var blockIDFlag BlockIDFlag
+	switch {
+	case vote.BlockID.IsComplete():
+		blockIDFlag = BlockIDFlagCommit
+		if len(vote.ExtensionSignature) == 0 {
+			panic(fmt.Sprintf("Invalid vote %v - a BlockID is complete but missing vote extension signature", vote))
```

Comment:
I hate that we just panic here, but I'm not sure there's much to be done.

Is this truly unrecoverable?

Reply:
I believe is it unrecoverable. If the vote extension signature is missing here, it is a malformed precommit vote. It's (at least) as serious a condition as line 130

Reply:
I have just checked the calling hierachy of this function. It's far from trivial, but there are two important callsites, namely when deciding a block, and when proposing a block, where we should NEVER find ourselves with a precommit `Vote` without an extension *signature*.
If that is the case, it is evidence that something very bad is going on in the consensus in-memory data structures. I think it is wrong to try to carry on with the execution in such circumstances.